Regression: jota.Transaction is not serializable.

https://github.com/iotaledger/compass/blob/612bc13b8e4e55850e3d043467c6b3b588f370ec/compass/conf/CoordinatorState.java#L40

replace with trytes themselves (Strings)
```
  public List<String> latestMilestoneTransactions;
```
```
      List<Transaction> latestMilestoneTransactions = db.createMilestone(trunk, branch, state.latestMilestoneIndex, config.MWM);
      state.latestMilestoneTransactions = latestMilestoneTransactions.stream().map(Transaction::toTrytes).collect(Collectors.toList());
      state.latestMilestoneHash = latestMilestoneTransactions.get(0).getHash();
```

```
main] INFO org.iota.compass.Coordinator - storing index 16
[main] ERROR org.iota.compass.Coordinator - Error saving Compass state to file '//compass.state'!
java.io.NotSerializableException: jota.model.Transaction
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at java.util.ArrayList.writeObject(ArrayList.java:766)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1140)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at org.iota.compass.Coordinator.storeState(Coordinator.java:101)
	at org.iota.compass.Coordinator.start(Coordinator.java:339)
	at org.iota.compass.Coordinator.main(Coordinator.java:136)
Exception in thread "main" java.lang.RuntimeException: java.io.NotSerializableException: jota.model.Transaction
	at org.iota.compass.Coordinator.start(Coordinator.java:344)
	at org.iota.compass.Coordinator.main(Coordinator.java:136)
Caused by: java.io.NotSerializableException: jota.model.Transaction
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1184)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at java.util.ArrayList.writeObject(ArrayList.java:766)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at java.io.ObjectStreamClass.invokeWriteObject(ObjectStreamClass.java:1140)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1496)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.defaultWriteFields(ObjectOutputStream.java:1548)
	at java.io.ObjectOutputStream.writeSerialData(ObjectOutputStream.java:1509)
	at java.io.ObjectOutputStream.writeOrdinaryObject(ObjectOutputStream.java:1432)
	at java.io.ObjectOutputStream.writeObject0(ObjectOutputStream.java:1178)
	at java.io.ObjectOutputStream.writeObject(ObjectOutputStream.java:348)
	at org.iota.compass.Coordinator.storeState(Coordinator.java:101)
	at org.iota.compass.Coordinator.start(Coordinator.java:339)
	... 1 more

```
the original error ^.

--- a/compass/Coordinator.java
+++ b/compass/Coordinator.java
@@ -80,27 +80,17 @@ public class Coordinator {
 
   private static CoordinatorState loadState() throws IOException, ClassNotFoundException {
     CoordinatorState state;
-    ObjectInputStream ois = null;
-    try {
-      ois = new ObjectInputStream(new FileInputStream(CoordinatorState.COORDINATOR_STATE_PATH));
+    try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(CoordinatorState.COORDINATOR_STATE_PATH))) {
       state = (CoordinatorState) ois.readObject();
-    } finally {
-      if (ois != null) {
-        ois.close();
-      }
+      log.info("loaded index {}", state.latestMilestoneIndex);
     }
     return state;
   }
 
   private void storeState(CoordinatorState state) throws IOException {
-    ObjectOutputStream oos = null;
-    try {
-      oos = new ObjectOutputStream(new FileOutputStream(CoordinatorState.COORDINATOR_STATE_PATH));
+    try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(CoordinatorState.COORDINATOR_STATE_PATH))) {
       oos.writeObject(state);
-    } finally {
-      if (oos != null) {
-        oos.close();
-      }
+      log.info("stored index {}", state.latestMilestoneIndex);
     }
   }
 
@@ -184,10 +174,10 @@ public class Coordinator {
 
   private void broadcastLatestMilestone() throws ArgumentException {
     if (config.broadcast) {
-      for (Transaction tx : state.latestMilestoneTransactions) {
-        api.storeAndBroadcast(tx.toTrytes());
+      for (String tx : state.latestMilestoneTransactions) {
+        api.storeAndBroadcast(tx);
       }
-      log.info("Broadcasted milestone: " + state.latestMilestoneIndex);
+      log.info("Broadcast milestone: " + state.latestMilestoneIndex);
     }
   }
 
@@ -229,7 +219,7 @@ public class Coordinator {
     log.info("Setting initial depth to: " + depth);
   }
 
-  private void start() throws ArgumentException, InterruptedException, IOException {
+  private void start() throws ArgumentException, InterruptedException {
     int bootstrap = config.bootstrap ? 0 : 3;
     int milestonePropagationRetries = 0;
     log.info("Bootstrap mode: " + bootstrap);
@@ -312,9 +302,10 @@ public class Coordinator {
 
       log.info("Issuing milestone: " + state.latestMilestoneIndex);
       log.info("Trunk: " + trunk + " Branch: " + branch);
-      state.latestMilestoneTransactions = db.createMilestone(trunk, branch, state.latestMilestoneIndex, config.MWM);
 
-      state.latestMilestoneHash = state.latestMilestoneTransactions.get(0).getHash();
+      List<Transaction> latestMilestoneTransactions = db.createMilestone(trunk, branch, state.latestMilestoneIndex, config.MWM);
+      state.latestMilestoneTransactions = latestMilestoneTransactions.stream().map(Transaction::toTrytes).collect(Collectors.toList());
+      state.latestMilestoneHash = latestMilestoneTransactions.get(0).getHash();
 
       // Do not store the state before broadcasting, since if broadcasting fails we should repeat the same milestone.
       broadcastLatestMilestone();

--- a/compass/Coordinator.java
+++ b/compass/Coordinator.java
@@ -105,7 +105,8 @@ public class Coordinator {
         .parse(args);
 
     // We want an empty state if bootstrapping
-    if (config.bootstrap) {
+      // and to allow overriding state file using `-index` flag
+    if (config.bootstrap || config.index != null) {
       state = new CoordinatorState();
     } else {
       try {

--- a/compass/Coordinator.java
+++ b/compass/Coordinator.java
@@ -45,7 +45,6 @@ import java.util.stream.Collectors;
 
 public class Coordinator {
   private static final Logger log = LoggerFactory.getLogger(Coordinator.class);
-  private static final String statePath = System.getProperty("user.dir") + File.separator + CoordinatorState.COORDINATOR_STATE_PATH;
   private final URL node;
   private final MilestoneSource db;
   private final IotaAPI api;
@@ -78,17 +77,17 @@ public class Coordinator {
     }).collect(Collectors.toList());
   }
 
-  private static CoordinatorState loadState() throws IOException, ClassNotFoundException {
+  private static CoordinatorState loadState(String path) throws IOException, ClassNotFoundException {
     CoordinatorState state;
-    try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(CoordinatorState.COORDINATOR_STATE_PATH))) {
+    try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream(path))) {
       state = (CoordinatorState) ois.readObject();
       log.info("loaded index {}", state.latestMilestoneIndex);
     }
     return state;
   }
 
-  private void storeState(CoordinatorState state) throws IOException {
-    try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(CoordinatorState.COORDINATOR_STATE_PATH))) {
+  private void storeState(CoordinatorState state, String path) throws IOException {
+    try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream(path))) {
       oos.writeObject(state);
       log.info("stored index {}", state.latestMilestoneIndex);
     }
@@ -110,9 +109,9 @@ public class Coordinator {
       state = new CoordinatorState();
     } else {
       try {
-        state = loadState();
+        state = loadState(config.statePath);
       } catch (Exception e) {
-        String msg = "Error loading Compass state file '" + statePath + "'! State file required if not bootstrapping...";
+        String msg = "Error loading Compass state file '" + config.statePath + "'! State file required if not bootstrapping...";
 
         log.error(msg, e);
         throw new RuntimeException(e);
@@ -325,9 +324,9 @@ public class Coordinator {
 
       // Everything went fine, now we store
       try {
-        storeState(state);
+        storeState(state, config.statePath);
       } catch (Exception e) {
-        String msg = "Error saving Compass state to file '" + statePath + "'!";
+        String msg = "Error saving Compass state to file '" + config.statePath + "'!";
 
         log.error(msg, e);
         throw new RuntimeException(e);

--- a/compass/Coordinator.java
+++ b/compass/Coordinator.java
@@ -97,6 +97,13 @@ public class Coordinator {
   public static void main(String[] args) throws Exception {
     CoordinatorConfiguration config = new CoordinatorConfiguration();
     CoordinatorState state;
+
+    JCommander.newBuilder()
+        .addObject(config)
+        .acceptUnknownOptions(true)
+        .build()
+        .parse(args);
+
     // We want an empty state if bootstrapping
     if (config.bootstrap) {
       state = new CoordinatorState();
@@ -104,18 +111,12 @@ public class Coordinator {
       try {
         state = loadState();
       } catch (Exception e) {
-        String msg = "Error loading Compass state file '" + statePath + "'! State file required if not bootstrapping.";
+        String msg = "Error loading Compass state file '" + statePath + "'! State file required if not bootstrapping...";
 
         log.error(msg, e);
         throw new RuntimeException(e);
       }
     }
-
-    JCommander.newBuilder()
-        .addObject(config)
-        .acceptUnknownOptions(true)
-        .build()
-        .parse(args);
 
     Coordinator coo = new Coordinator(config, state, SignatureSourceHelper.signatureSourceFromArgs(config.signatureSource, args));
     coo.setup();

--- a/compass/conf/CoordinatorConfiguration.java
+++ b/compass/conf/CoordinatorConfiguration.java
@@ -58,4 +58,7 @@ public class CoordinatorConfiguration extends BaseConfiguration {
   @Parameter(names = "-propagationRetriesThreshold", description = "Number of milestone propagation retries we attempt before failing.")
   public int propagationRetriesThreshold = 5;
 
+  @Parameter(names = "-statePath", description = "Path to compass state file.")
+  public String statePath = "compass.state";
+
 }

--- a/compass/conf/CoordinatorState.java
+++ b/compass/conf/CoordinatorState.java
@@ -31,8 +31,6 @@ import java.util.List;
 
 public class CoordinatorState implements Serializable {
 
-  public static final String COORDINATOR_STATE_PATH = "compass.state";
-
   public int latestMilestoneIndex;
   public String latestMilestoneHash;
   public long latestMilestoneTime;

--- a/compass/conf/CoordinatorState.java
+++ b/compass/conf/CoordinatorState.java
@@ -25,9 +25,8 @@
 
 package org.iota.compass.conf;
 
-import jota.model.Transaction;
-
 import java.io.Serializable;
+import java.util.Collections;
 import java.util.List;
 
 public class CoordinatorState implements Serializable {
@@ -37,6 +36,6 @@ public class CoordinatorState implements Serializable {
   public int latestMilestoneIndex;
   public String latestMilestoneHash;
   public long latestMilestoneTime;
-  public List<Transaction> latestMilestoneTransactions;
+  public List<String> latestMilestoneTransactions = Collections.EMPTY_LIST;
 
 }

